Handle Tern server and interface re-initialization on node restart

swmitra opened the pull request:
Adding logic to re-initialize Tern module on node restart. 
@petetnt @zaggino @nethip @saurabh95  Can you please review this PR?

swmitra committed:
Handle Tern server and interface re-initialization on node restart

petetnt commented on src/extensions/default/JavaScriptCodeHints/ScopeManager.js:
```diff
@@ -1086,23 +1099,20 @@ define(function (require, exports, module) {
                         handleTimedOut(response);
                     } else if (type === MessageIds.TERN_WORKER_READY) {
                         moduleDeferred.resolveWith(null, [_ternNodeDomain]);
+                    } else if (type === "ReInitTern") {
```
This should be a const like the other ones (`MessageIds.RE_INIT_TERN`)

petetnt replied:
Wait, I just saw that this only gets called if `!messageIds`, so maybe just `RE_INIT_TERN`

swmitra replied:
Fixed

petetnt commented on src/extensions/default/JavaScriptCodeHints/ScopeManager.js:
```diff
@@ -1086,23 +1099,20 @@ define(function (require, exports, module) {
                     } else if (type === MessageIds.TERN_WORKER_READY) {
                         moduleDeferred.resolveWith(null, [_ternNodeDomain]);
+                    } else if (type === "ReInitTern") {
+                        // Ensure the request is beacuse of a node restart
```
Nit typo: `because`

swmitra replied:
Fixed

petetnt commented on src/extensions/default/JavaScriptCodeHints/ScopeManager.js:
```diff
@@ -1334,7 +1344,7 @@ define(function (require, exports, module) {
             // because it's easier to debug if the module isn't
             // getting reset all the time.
-            if (force || (!config.noReset && ++_hintCount > MAX_HINTS)) {
+            if (currentModule.resetForced || force || (!config.noReset && ++_hintCount > MAX_HINTS)) {
```
Should `currentModule.resetForced` be set back to `false` at some point? 🤔

swmitra replied:
`currentModule` is reassigned with `newTernModule` soon after ...

petetnt replied:
Thanks for confirming 👍

petetnt commented on src/extensions/default/JavaScriptCodeHints/node/TernNodeDomain.js:
```diff
@@ -830,6 +841,7 @@ function init(domainManager) {
             }
         ]
     );
+    setTimeout(checkInterfaceAndReInit, 5000);
```
Any specific reason for the `5000` timeout or is it just in case?

swmitra replied:
I can reduce it to 1000 as well

petetnt replied:
Are there any heuristics that could help one figure out when it's a good time to reinit?

swmitra replied:
I was trying to find it out, but it's really dependent on environment. It's basically the time to re-establish NodeConnections when a node restart has happened. Tested on 3 different environments and it seems really fast. I have changed the timer duration to 1s now.

petetnt replied:
🆗  any idea what happens if the the `NodeConnections` aren't up by that time? Hints just don't work?

petetnt reviewed (commented):
Some initial comments

swmitra committed:
Address review comments

swmitra commented:
@petetnt As usual, it was really quick 💯 
I have addressed the review comments, you can check it once more. 
I have tested it by forcefully killing the node process and after node restart trying to get code hints. For a brief period of time (few seconds) there won't be any hints. But that's better than not having hints at all after a node restart.

nethip commented:
@swmitra Please unit test these changes on Linux as well as Mac.

petetnt reviewed (approved):
Code LGTM
